src/main/java/de/tum/cit/aet/artemis/exercise/service/ParticipationFilterService.java

```diff
@@ -0,0 +1,117 @@
+     * at most one participation. For programming exercises, an additional practice participation can exist.
+     *
+     * @param participationsAcrossAllExercises all participations of the student in all exercises of the course
+     * @param exercise                         the exercise for which we want the participations
+     * @return the participations in the exercise
+     */
+    public Set<StudentParticipation> findStudentParticipationsInExercise(Set<StudentParticipation> participationsAcrossAllExercises, Exercise exercise) {
+        // only consider participations in the given exercise
+        if (participationsAcrossAllExercises == null || participationsAcrossAllExercises.isEmpty()) {
+            return Set.of();
+        }
+        var participationsInExercise = participationsAcrossAllExercises.stream().filter(p -> p.getExercise() != null && p.getExercise().equals(exercise))
+                .collect(Collectors.toSet());
+
+        if (participationsInExercise.isEmpty()) {
+            return Set.of();
+        }
+
+        if (ExerciseType.PROGRAMMING.equals(exercise.getExerciseType())) {
+            return findStudentParticipationsForProgrammingExercises(participationsInExercise);
+        }
+        else {
+            return Set.of(findStudentParticipationForNonProgrammingExercises(participationsInExercise));
+        }
+    }
+
+    /**
+     * Filters the student participation for display in the student dashboard.
+     *
+     * @param participation the participation that contains the exercise and should be filtered
+     * @param isStudent     used to determine if further filtering is needed.
+     */
+    public void filterParticipationForCourseDashboard(StudentParticipation participation, boolean isStudent) {
+        Optional<Submission> optionalSubmission = submissionFilterService.getLatestSubmissionWithResult(participation.getSubmissions(), false);
+
+        Set<Result> results = Set.of();
+
+        if (optionalSubmission.isPresent()) {
+            var submission = optionalSubmission.get();
+            var latestResult = submission.getLatestResult();
+            if (latestResult != null) {
+                results = Set.of(latestResult);
+                latestResult.setParticipation(null);
+                // filter sensitive information about the assessor if the current user is a student
+                if (isStudent) {
+                    latestResult.filterSensitiveInformation();
+                }
+            }
+            submission.setResults(new ArrayList<>(results));
+        }
+
+        // add submission to participation or set it to null
+        participation.setSubmissions(optionalSubmission.map(Set::of).orElse(null));
+        participation.setResults(results);
+        if (optionalSubmission.isPresent()) {
+            optionalSubmission.get().setResults(new ArrayList<>(results));
+        }
```

Comment:
```suggestion
        optionalSubmission.ifPresent(submission -> submission.setResults(new ArrayList<>(results)));
```
nitpick/preference to avoid branching